### PR TITLE
V3 sheet functions

### PR DESCRIPTION
**Pull request by naterush**
# Description

Adds:
1. AND
2. AVG
3. CONCAT
4. MAX
5. MIN
6. MULTIPLY
7. OR
8. SUM

Does not add: STDEV, VAR, CORR, KURT, SKEW because it feels like a bad use of time right now. I feel like there are much higher priorities for me to spend 5 hours on - namely, other much more popular functions. 

# Testing

See tests. Add tests.

# Documentation

No.

### CODE REVIEW

File: mitosheet/mitosheet/public/v3/sheet_functions/bool_functions.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    }
+    """
+
+    # If we don't find any arguements, we default to True. Excel in practice
+    # defaults to a value error if there are no args, but it's ok
+    return get_final_result_series_or_primitive(
+        True,
+        argv,
+        lambda df: df.all().all(),
+        # Note the handling we do here to ensure that None's don't get propagated forward. We fill None
```
Comment (naterush) [inline mitosheet/mitosheet/public/v3/sheet_functions/bool_functions.py:57]: Remove

**Review by naterush — COMMENTED** [review]
Cleanup!

File: mitosheet/mitosheet/tests/public/v3/sheet_functions/bool_functions/test_and.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    # Series
+    ([True, 'abc', pd.Series([True,True,False])], pd.Series([True, True, False])),
+    (['abc', pd.Series([True,True,False])], pd.Series([True, True, False])),
+    ([True, pd.Series(['abc',True, False])], pd.Series([True, True, False])),
+
+    # Dataframes
+    ([pd.DataFrame({'a': [True, True, True], 'b': [True, True, 'abc']}), pd.Series([True,True,False])], pd.Series([True,True,False])),
+    ([pd.DataFrame({'a': [True, True, True], 'b': [True, True, 'abc']}), pd.Series([False, True,'abc'])], pd.Series([False,True,True])),
+
+    # Rolling ranges
```
Comment (aarondr77) [inline mitosheet/mitosheet/tests/public/v3/sheet_functions/bool_functions/test_and.py:53]: Making sure we test each of these is great. I would also be down to do one test for each spreadsheet function that has one arg of each type

File: mitosheet/mitosheet/tests/public/v3/sheet_functions/number_functions/test_sum.py [hunk 3]
```diff
@@ -77,6 +77,7 @@
 @pytest.mark.parametrize("_argv,expected", SUM_VALID_TESTS + SUM_INVALID_CAST_TESTS)
 def test_sum(_argv, expected):
     result = SUM(*_argv)
+    print(result)
```
Comment (aarondr77) [inline mitosheet/mitosheet/tests/public/v3/sheet_functions/number_functions/test_sum.py:80]: remove

File: mitosheet/mitosheet/public/v3/sheet_functions/utils.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    2. They update the result with each arg in two steps, preprocessing the arg and then combining that with the result
+    3. Preprocessing the arg:
+        - For dataframe values, they turned into primitive values with get_primitive_value_from_dataframe
+        - For rolling ranges, they are turned into series using repeated application of get_primitive_value_from_dataframe
+    4. Combining with the previous result. We are either combining two primtiive values, a primitive value and a series, or two series
+        - If combining two primitive values, we combine with get_new_result_from_primitive_values
+        - If combining a primitive value and a series, we use a .apply on the series with get_new_result_from_primitive_values
+        - If combining two series, then we use get_new_result_from_series
+
+    This makes it much easier to specify many functions like AND, SUM, OR, etc.
```
Comment (aarondr77) [inline mitosheet/mitosheet/public/v3/sheet_functions/utils.py:122]: Yeah this is an awesome helper function! Nice work!

File: mitosheet/mitosheet/public/v3/sheet_functions/bool_functions.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    }
+    """
+
+    # If we don't find any arguements, we default to True. Excel in practice
+    # defaults to a value error if there are no args
+    return get_final_result_series_or_primitive(
+        False,
+        argv,
+        lambda df: df.any().any(),
+        lambda previous_value, new_value: (previous_value or new_value),
```
Comment (aarondr77) [inline mitosheet/mitosheet/public/v3/sheet_functions/bool_functions.py:92]:
```suggestion
        lambda previous_value, new_value: previous_value or new_value,
```

File: mitosheet/mitosheet/public/v3/sheet_functions/number_functions.py [hunk 6]
```diff
@@ -57,62 +58,65 @@ def AVG(*argv: Union[int, float, None, pd.Series, RollingRange]) -> NumberFuncti
...
+        lambda previous_value, new_value: max(previous_value, new_value),
+        lambda previous_series, new_series: pd.concat([previous_series, new_series], axis=1).max(axis=1)
+    )
 
-    for arg in argv:
-        if isinstance(arg, pd.DataFrame):
-            df_sum = arg.sum().sum()
-            result = get_new_result(result, df_sum)
+    # If we don't find any arguements, we default to 0 -- like Excel
+    kept_default_max_value = not isinstance(result, pd.Series) and result == (-sys.maxsize - 1)
```
Comment (aarondr77) [inline mitosheet/mitosheet/public/v3/sheet_functions/number_functions.py:70]: Do we need this first check `not isinstance(result, pd.Series)`?

File: mitosheet/mitosheet/public/v3/sheet_functions/number_functions.py [hunk 7]
```diff
@@ -57,62 +58,65 @@ def AVG(*argv: Union[int, float, None, pd.Series, RollingRange]) -> NumberFuncti
...
-            result = get_new_result(result, arg)
-            
-        elif arg is not None:
-            result = get_new_result(result, arg)
+
+@cast_values_in_arg_to_type('number')
+def MIN(*argv: Union[int, float, None, pd.Series, RollingRange]) -> NumberFunctionReturnType:
+
+    result = get_final_result_series_or_primitive(
+        sys.maxsize,
```
Comment (aarondr77) [inline mitosheet/mitosheet/public/v3/sheet_functions/number_functions.py:78]: Not sure if it matters, but for the MAX function we used `-sys.maxsize - 1`. Do we want to use `sys.maxsize - 1` here?

Comment (aarondr77) [reply mitosheet/mitosheet/public/v3/sheet_functions/number_functions.py:78]: Maybe we could create a MAX_SIZE constant that we use in all these cases. Sometimes -MAX_SIZE, other times MAX_SIZE

File: mitosheet/mitosheet/public/v3/sheet_functions/string_functions.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+                "description": "Additional strings or series to append in sequence."
+            }
+        ]
+    }
+    """
+
+    return get_final_result_series_or_primitive(
+        '',
+        argv,
+        lambda df: df.sum().sum(), # Concats in the same order as Excel - yay!
```
Comment (aarondr77) [inline mitosheet/mitosheet/public/v3/sheet_functions/string_functions.py:48]: Nice!

File: mitosheet/mitosheet/tests/public/v3/sheet_functions/string_functions/test_concat.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+from mitosheet.public.v3.sheet_functions.string_functions import CONCAT
+
+
+CONCAT_VALID_TESTS: Any = [
+    # Just constant tests
+    (['a', 'b'], 'ab'),
+    (['a', 1, True], 'a1True'),
+
+    # Constants and series
+    (['a', pd.Series(['a', 'b', 'c'])], pd.Series(['aa', 'ab', 'ac'])),
```
Comment (aarondr77) [inline mitosheet/mitosheet/tests/public/v3/sheet_functions/string_functions/test_concat.py:23]: Can we add a `None` test here? It probably makes sense to make sure we have at least one None test per function.

**Review by aarondr77 — COMMENTED** [review]
Code looks great. I love when it turns out so clean like this! Great work. 

No major changes here, just a few random cleanup.

File: mitosheet/mitosheet/public/v3/sheet_functions/utils.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+from mitosheet.public.v3.rolling_range import RollingRange
+
+
+PrimitiveType = TypeVar('PrimitiveType', bound=Union[str, float, int, bool, datetime, timedelta])
+
+ResultType = Union[pd.Series, PrimitiveType]
+
+
+def __get_default_value_if_value_is_none_or_nan(value: PrimitiveType, default_value: PrimitiveType) -> PrimitiveType:
+    if value is None or (isinstance(value, float) and np.isnan(value)):
```
[the remark this answers is not in the document]
Comment (naterush) [reply mitosheet/mitosheet/public/v3/sheet_functions/utils.py:19]: Because all NaN values are float values. You can't just put any old object into np.isnan or it will error. Something can only be an NaN if it's a float (or... if it's a date time - but we'll address that when we get there!).

File: mitosheet/mitosheet/public/v3/sheet_functions/number_functions.py (hunk 7 above)
Comment (naterush) [reply mitosheet/mitosheet/public/v3/sheet_functions/number_functions.py:78]: You do minus one because that's what the maximum negative value is (they aren't symmetric). So defining a single variable wouldn't work in this case.

File: mitosheet/mitosheet/public/v3/sheet_functions/number_functions.py (hunk 6 above)
Comment (naterush) [reply mitosheet/mitosheet/public/v3/sheet_functions/number_functions.py:70]: Yes - because pandas throws a warning. If you check equality with a series, you get a series of booleans - which isn't what we want!

File: mitosheet/mitosheet/public/v3/sheet_functions/utils.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+
+from mitosheet.public.v3.rolling_range import RollingRange
+
+
+PrimitiveType = TypeVar('PrimitiveType', bound=Union[str, float, int, bool, datetime, timedelta])
+
+ResultType = Union[pd.Series, PrimitiveType]
+
+
+def __get_default_value_if_value_is_none_or_nan(value: PrimitiveType, default_value: PrimitiveType) -> PrimitiveType:
```
[the remark this answers is not in the document]
Comment (naterush) [reply mitosheet/mitosheet/public/v3/sheet_functions/utils.py:18]: I don't want anyone using them outside the file. Double underscore stops them from being exported, I think!

File: mitosheet/mitosheet/public/v3/sheet_functions/bool_functions.py [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+            },
+            {
+                "element": "boolean_condition2 ... [OPTIONAL]",
+                "description": "An expression or series that returns True or False values. See IF documentation for a list of conditons."
+            }
+        ]
+    }
+    """
+
+    # If we don't find any arguements, we default to True. Excel in practice
```
[the remark this answers is not in the document]
Comment (naterush) [reply mitosheet/mitosheet/public/v3/sheet_functions/bool_functions.py:51]: This breaks the default value filling. I think this is ok.

File: mitosheet/mitosheet/public/v3/sheet_functions/bool_functions.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+def AND(*argv: Union[bool, None, pd.Series, RollingRange]) -> BoolFunctionReturnType:
+    """
+    {
+        "function": "AND",
+        "description": "Returns True if all of the provided arguments are True, and False if any of the provided arguments are False.",
+        "search_terms": ["and", "&", "if", "conditional"],
+        "examples": [
+            "AND(True, False)",
+            "AND(Nums > 100, Nums < 200)",
+            "AND(Pay > 10, Pay < 20, Status == 'active')"
```
[the remark this answers is not in the document]
Comment (naterush) [reply mitosheet/mitosheet/public/v3/sheet_functions/bool_functions.py:36]: Would be down to leave a documentation pass till the very end!
